source/common/config/version_converter.cc

```diff
@@ -65,87 +57,165 @@ void constMessageFieldVisitor(
+      const std::string& original_type = unknown_field.length_delimited();
+      const Protobuf::Descriptor* original_descriptor =
+          Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(original_type);
+      auto result = createForDescriptorWithCast(upgraded_message, original_descriptor);
+      // We should clear out the OriginalTypeFieldNumber in the recovered message.
+      eraseOriginalTypeInformation(*result->msg_);
+      return result;
+    }
+  }
+  return createForDescriptorWithCast(upgraded_message, nullptr);
 }
 
 DynamicMessagePtr VersionConverter::downgrade(const Protobuf::Message& message) {
-  auto downgraded_message = std::make_unique<DynamicMessage>();
   const Protobuf::Descriptor* prev_desc = ApiTypeOracle::getEarlierVersionDescriptor(message);
-  if (prev_desc != nullptr) {
-    downgraded_message->msg_.reset(
-        downgraded_message->dynamic_msg_factory_.GetPrototype(prev_desc)->New());
-    wireCast(message, *downgraded_message->msg_);
-    return downgraded_message;
-  }
-  // Unnecessary copy, since the existing message is being treated as
-  // "downgraded". However, we want to transfer an owned object, so this is the
-  // best we can do.
-  const Protobuf::Descriptor* desc = message.GetDescriptor();
-  downgraded_message->msg_.reset(
-      downgraded_message->dynamic_msg_factory_.GetPrototype(desc)->New());
-  downgraded_message->msg_->MergeFrom(message);
-  return downgraded_message;
+  return createForDescriptorWithCast(message, prev_desc);
 }
 
 DynamicMessagePtr
 VersionConverter::reinterpret(const Protobuf::Message& message,
                               envoy::config::core::v3alpha::ApiVersion api_version) {
   switch (api_version) {
   case envoy::config::core::v3alpha::ApiVersion::AUTO:
-  case envoy::config::core::v3alpha::ApiVersion::V2:
+  case envoy::config::core::v3alpha::ApiVersion::V2: {
     // TODO(htuch): this works as long as there are no new fields in the v3+
     // DiscoveryRequest. When they are added, we need to do a full v2 conversion
     // and also discard unknown fields. Tracked at
     // https://github.com/envoyproxy/envoy/issues/9619.
-    return downgrade(message);
+    auto dynamic_message = downgrade(message);
+    eraseOriginalTypeInformation(*dynamic_message->msg_);
+    return dynamic_message;
+  }
   case envoy::config::core::v3alpha::ApiVersion::V3ALPHA: {
     // We need to scrub the hidden fields.
     auto non_shadowed_message = std::make_unique<DynamicMessage>();
-    const Protobuf::Descriptor* desc = message.GetDescriptor();
-    non_shadowed_message->msg_.reset(
-        non_shadowed_message->dynamic_msg_factory_.GetPrototype(desc)->New());
+    non_shadowed_message->msg_.reset(message.New());
     non_shadowed_message->msg_->MergeFrom(message);
     VersionUtil::scrubHiddenEnvoyDeprecated(*non_shadowed_message->msg_);
+    eraseOriginalTypeInformation(*non_shadowed_message->msg_);
```

Comment:
I'm slightly worried about forgetting to do the various transforms in various scenarios. Are they all localized to this file? Is there any benefit in having these all be bundled into the transform/cast function with enums on what additional things to do? TIOLI just throwing it out there.

Reply:
I think broadly the transformations depend on the macro-level operation.e.g. are we reinterpreting vs. upgrading vs. downgrading, so are mostly local. There are some exceptions you can see in this PR outside of test:
1. Local info `Node` type erasure. We synthesize messages for various discovery services and include `Node`, which comes from the wire with original type information added during upgrades. I had in the PR that you reviewed the erasure taking place in `local_info_impl.h`. I think the right thing to do is make sure things like load stats reporter, HDS, etc. use `VersionConverter::reinterpret`, so I will fix to do this and get rid of that exception case.
2. `eds.cc` uses this to scrub on a best effort for reducing per-host overhead of annotations.